### DISCUSSION TITLE
Documentation for writing key bindings from scratch

### DISCUSSION THREAD

**Discussion by brabalan** [meow-edit/meow#555] [Q&A] [answered]
I use a custom keyboard layout, and as I am discovering meow, I want to write the key bindings from scratch. I unfortunately cannot find documentation about this, with the exception of the examples and the tutorial. At the moment, I am following the tutorial and adding key bindings as concepts are introduced, but it could be very useful to have some information about them.

For instance, what I would like is
- a list of the functions to bind
- some suggestions for the keys (I have for instance discovered that `n` and `p` move down and up in many modes, so they may be good movement keys)
- an explanation of how to write `meow-cheatsheet-layout` and why it matters
- an explanation of what `meow-motion-overwrite-define-key` does

Is this information already described somewhere?

**Comment by DogLooksGood** [comment, marked as answer]
Hey,

I believe the example in [Get Started](https://github.com/meow-edit/meow/blob/master/GET_STARTED.org) can be where you start with. 
 It shows you all the useful functions to bind. Also there's an issue #506 discussing a ergonomic key binding design. 

I do have some principles in mind for how a good command layout should look like. For example, the commands should be arranged by their names mostly yet still be ergonomic and efficient. And to be efficient, we do not only care about the position of each command, but need all high-frequency used key combinations to be easy to press. Personally, I have no wonderful solution so far, especially for all keyboard layouts.

`meow-cheatsheet-layout` is useful when you are using a custom keyboard layout, and want to good looking help page.  I do not recommend to spend too much time on it as start.

Check [here](https://github.com/meow-edit/meow/blob/master/TUTORIAL.org#motion) for key bindings in motion state.

**Comment by brabalan** [comment]
Thank you for the link for the bindings in motion state, it really helps.

I have a followup about mode-specific keymaps. I have defined the following for notmuch:
```emacs-lisp
(use-package notmuch
  :bind (:map notmuch-show-mode-map
	      ("x" . notmuch-show-archive-message-then-next-or-next-thread)
	      ("d" . as/notmuch/show-delete)))
```

and it works well, because `x` and `d` are free keys. If I replace `x` by `a`, I get an insert command instead of the one I bound. Is there a way to make sure the mode keymap has priority? Or is it a question of when I run my `meow-setup`?

**Reply by DogLooksGood** [reply]
Which mode is enabled in your case? In a non-editing buffer, the motion mode should be enabled.

**Reply by brabalan** [reply]
This is what emacs tells me:
```
Minor modes enabled in this buffer: Font-Lock Meow Meow-Normal
Visual-Line

The major mode is notmuch-show mode defined in notmuch-show.el:
```

**Reply by DogLooksGood** [reply]
Try this one.

```
(add-to-list 'meow-mode-state-list '(notmuch-show-mode . motion))
```

**Reply by brabalan** [reply]
Ah, that was it. And I understand better that mode now (thanks to the link you pointed to earlier).
